blockdag/utxoset.go

```diff
@@ -166,6 +180,71 @@ func NewUTXODiff() *UTXODiff {
 	}
 }
 
+// checkIntersection checks if there is an intersection between two utxoCollections
+func checkIntersection(uc1 utxoCollection, uc2 utxoCollection) bool {
+	for outpoint := range uc1 {
+		if uc2.contains(outpoint) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// checkIntersectionWithRule checks if there is an intersection between two utxoCollections satisfying arbitrary rule
+func checkIntersectionWithRule(uc1 utxoCollection, uc2 utxoCollection, extraRule func(wire.Outpoint, *UTXOEntry, *UTXOEntry) bool) bool {
+	for outpoint, utxoEntry := range uc1 {
+		if diffEntry, ok := uc2.get(outpoint); ok {
+			if extraRule(outpoint, utxoEntry, diffEntry) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// intersectionWithRemainder calculates an intersection between two utxoCollections, puts it into result
+// and returns the remainder
+func intersectionWithRemainder(result, uc1, uc2 utxoCollection) utxoCollection {
```

Comment:
Out parameters were used for the possibility to perform in-place operations if we need that later.
But I see that these could be quite specific operations and perhaps this part adds some code complexity.
I will use multiple-return as you suggest.

Reply:
After benchmarking the multiple-return version, I noticed a significant performance drop. I did some optimizations and end up having mostly the same code as I initially had - i.e. modifying result in-place while performing set operations and making some preallocations for return values.

These parts are crucial as they:
- save a huge amount of time for preallocations;
- improves the readability of "diffFrom" function for a little bit (that's a value judgement, of course).

And as we are not going (at least at that time) to use these set-related operations as a general-purpose set operations (and we shouldn't as these are for utxoCollection specific type), I believe that's a right decision to keep these parts at least for the sake of performance.
IMO, perhaps the only thing that hurts because of the in-place operations is the thing that it's not that Go-ish approach.